tasks/ast_codegen/src/generators/derive_clone_in.rs

```diff
@@ -0,0 +1,134 @@
+
+        let header = generated_header!();
+
+        GeneratorOutput::Stream((
+            output(crate::AST_CRATE, "derive_clone_in.rs"),
+            quote! {
+                #header
+
+                use oxc_allocator::{Allocator, CloneIn};
+                use crate::ast::*;
+                endl!();
+
+                #(#impls)*
+            },
+        ))
+    }
+}
+
+fn derive_enum(def: &EnumDef) -> TokenStream {
+    let ty_ident = def.ident();
+    impl_clone_in(&ty_ident, def.has_lifetime, || {
+        let mut used_alloc = false;
+        let matches = def
+            .all_variants()
+            .map(|var| {
+                let ident = var.ident();
+                if var.is_unit() {
+                    quote!(Self :: #ident => Self :: Cloned :: #ident)
+                } else {
+                    used_alloc = true;
+                    quote!(Self :: #ident(it) => Self :: Cloned :: #ident(it.clone_in(alloc)))
+                }
+            })
+            .collect_vec();
+        let alloc_ident = if used_alloc { format_ident!("alloc") } else { format_ident!("_") };
+        (
+            alloc_ident,
+            quote! {
+                match self {
+                    #(#matches),*
+                }
+            },
+        )
+    })
+}
+
+fn derive_struct(def: &StructDef) -> TokenStream {
+    let ty_ident = def.ident();
+    impl_clone_in(&ty_ident, def.has_lifetime, || {
+        let (alloc_ident, body) = if def.fields.is_empty() {
+            (format_ident!("_"), TokenStream::default())
+        } else {
+            let fields = def.fields.iter().map(|field| {
+                let ident = field.ident();
+                quote!(#ident: self.#ident.clone_in(alloc))
+            });
+            (format_ident!("alloc"), quote!({ #(#fields),* }))
+        };
+        (
+            alloc_ident,
+            quote! {
+                    #ty_ident #body
+            },
+        )
+    })
+}
+
+fn impl_clone_in<F>(ty_ident: &Ident, has_lifetime: bool, body: F) -> TokenStream
+where
+    F: FnOnce() -> (/* allocator name */ Ident, TokenStream),
+{
+    let (alloc, body) = body();
+    let (impl_lifetimes, trait_lifetime, old_lifetime, new_lifetime, alloc_ref) = if has_lifetime {
+        (
+            quote!(<'old_alloc, 'new_alloc>),
+            quote!(<'new_alloc>),
+            quote!(<'old_alloc>),
+            quote!(<'new_alloc>),
+            quote!(&'new_alloc),
+        )
+    } else {
+        (
+            quote!(<'alloc>),
+            quote!(<'alloc>),
+            TokenStream::default(),
+            TokenStream::default(),
+            quote!(&'alloc),
+        )
+    };
+
+    quote! {
+        endl!();
+        impl #impl_lifetimes CloneIn #trait_lifetime for #ty_ident #old_lifetime {
+            type Cloned = #ty_ident #new_lifetime;
+
+            fn clone_in(&self, #alloc: #alloc_ref Allocator) -> Self::Cloned {
+                #body
+            }
+        }
+    }
+}
```

Comment:
This code is very DRY, but personally I feel that makes it harder to understand. How about this (more repetitive but clearer):

```rs
if has_lifetime {
    quote! {
        impl <'old_alloc, 'new_alloc> CloneIn<'new_alloc> for #ty_ident<'old_alloc> {
            type Cloned = #ty_ident<'new_alloc>;
            fn clone_in(&self, #alloc: &'new_alloc Allocator) -> Self::Cloned {
                #body
            }
        }
    }
} else {
    quote! {
        impl <'new_alloc> CloneIn<'new_alloc> for #ty_ident {
            type Cloned = #ty_ident;
            fn clone_in(&self, #alloc: &'new_alloc Allocator) -> Self::Cloned {
                #body
            }
        }
    }
}
```

Also: Instead of taking a closure, can't this function just take `alloc: Ident, body: TokenStream` params instead? I don't see what value the closure adds.

Reply:
I initially was doing most of the checks in the `impl_clone_in` function. Now that I've refactored it it doesn't need the body to be a closure.

I'll submit a patch for it.